packages/access-api/src/models/delegations.js

```diff
@@ -179,3 +191,169 @@ export function delegationsTableBytesToArrayBuffer(sqlValue) {
     return Uint8Array.from(sqlValue)
   }
 }
+
+export const delegationsV3Table = /** @type {const} */ (`delegations_v3`)
+
+/**
+ * @template {Database<import('../types/access-api-cf-db').DelegationsV3Tables>} DB
+ */
+export class DbDelegationsStorageWithR2 {
+  // @todo abstract away R2 specifics into DagStore: ~AsyncMap<CID, Ucanto.Delegation>
+  /** @type {R2Bucket} */
+  #dags
+  /** @type {DB} */
+  #db
+  #delegationsTableName = delegationsV3Table
+  /* @type {(d: { cid: string }) => string} */
+  #getDagsKey = carFileKeyer
+
+  /**
+   * @param {DB} db
+   * @param {R2Bucket} dags
+   */
+  constructor(db, dags) {
+    this.#db = db
+    this.#dags = dags
+    // eslint-disable-next-line no-void
+    void (
+      /** @type {import('../types/delegations').DelegationsStorage} */ (this)
+    )
+  }
+
+  /**
+   * store items
+   *
+   * @param  {Array<Ucanto.Delegation>} delegations
+   * @returns {Promise<void>}
+   */
+  async putMany(...delegations) {
+    if (delegations.length === 0) {
+      return
+    }
+    await writeDelegations(this.#dags, delegations, this.#getDagsKey)
+    const values = delegations.map((d) => createDelegationRowUpdateV3(d))
+    // @todo - if this fails, undo writeDelegations that dont need to be stored
+    await this.#db
+      .insertInto(this.#delegationsTableName)
+      .values(values)
+      .onConflict((oc) => oc.column('cid').doNothing())
+      .executeTakeFirst()
+  }
+
+  /** @returns {Promise<bigint>} */
+  async count() {
+    return count(this.#db, this.#delegationsTableName)
+  }
+
+  /**
+   * @param {import('../types/delegations').Query} query
+   */
+  async *find(query) {
+    const { audience } = query
+    const delegations = this.#delegationsTableName
+    const selection = await this.#db
+      .selectFrom(delegations)
+      .selectAll()
+      .where(`${delegations}.audience`, '=', audience)
+      .execute()
+    for await (const row of selection) {
+      yield this.#rowToDelegation(row)
```

Comment:
Nit: Should we perhaps not throw on first error and try to yield as much as we can and then throw ?

Reply:
I'd prefer to fail fast. afaict there is no good reason to keep going if this errors. something is very wrong